File: clients/go/debugging/colored_vertex.go
package debugging

import (
	. "aicup22/model"
	"io"
)

// Point + color
type ColoredVertex struct {
	// Position
	Position Vec2
	// Color
	Color Color
}

func NewColoredVertex(position Vec2, color Color) ColoredVertex {
	return ColoredVertex{
		Position: position,
		Color:    color,
	}
}

// Read ColoredVertex from reader
func ReadColoredVertex(reader io.Reader) ColoredVertex {
	position := ReadVec2(reader)
	color := ReadColor(reader)
	return ColoredVertex{
		Position: position,
		Color:    color,
	}
}

// Write ColoredVertex to writer
func (coloredVertex ColoredVertex) Write(writer io.Writer) {
	position := coloredVertex.Position
	position.Write(writer)
	color := coloredVertex.Color
	color.Write(writer)
}

// Get string representation of ColoredVertex
func (coloredVertex ColoredVertex) String() string {
	stringResult := "{ "
	stringResult += "Position: "
	position := coloredVertex.Position
	stringResult += position.String()
	stringResult += ", "
	stringResult += "Color: "
	color := coloredVertex.Color
	stringResult += color.String()
	stringResult += " }"
	return stringResult
}

File: clients/rust/src/debugging/colored_vertex.rs
use super::*;

/// Point + color
#[derive(Clone, Debug)]
pub struct ColoredVertex {
    /// Position
    pub position: model::Vec2,
    /// Color
    pub color: debugging::Color,
}

impl trans::Trans for ColoredVertex {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.position.write_to(writer)?;
        self.color.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let position: model::Vec2 = trans::Trans::read_from(reader)?;
        let color: debugging::Color = trans::Trans::read_from(reader)?;
        Ok(Self {
            position,
            color,
        })
    }
}